gunrock/cuda/launch_box.hxx

```diff
@@ -12,44 +12,148 @@
+  #define SM_TAG   sm_35
+  #define ARCH_TAG kepler
+#elif __CUDA_ARCH__ == 320
+  #define SM_TAG   sm_32
+  #define ARCH_TAG kepler
+#elif __CUDA_ARCH__ >= 300
+  #define SM_TAG   sm_30
+  #define ARCH_TAG kepler
+#else
+  #error "Gunrock only supports sm_30 and above"
+#endif  // __CUDA_ARCH__
+#else
+  // What should these macros be on the host side?
+  #define SM_TAG   sm_00
+  #define ARCH_TAG ampere
+#endif
+
 namespace gunrock {
 namespace cuda {
 
-// Should the struct members be template parameters?
-template<int blockDim_, int gridDim_, size_t smemBytes_ = 0>
-struct launch_params_t{
-  int blockDim = blockDim_;
-  int gridDim = gridDim_;
-  size_t smemBytes = smemBytes_;
-};
+//////////////// Move this section to another file?
+/**
+ * @brief Blank struct to use as a default base to make inheritance optional
+ */
+struct empty_t {};
 
-// template<int ccCombined, int... launch_tparams>
-// struct cc_launch_params_t : launch_params_t<launch_tparams...> {
-//     // Compute capability related things
-// };
+// Inheritance structs to expand the variadic launch types passed into the launch box and define each them within the launch box's namespace
+template<typename... base_v>
+struct inherit_t;
 
-// smemBytes_ isn't an int, but I am assuming it will get cast to a size_t? I think there is a better way to generalize the type
-template<int... launch_tparams>
-struct arch_launch_params_t : launch_params_t<launch_tparams...> {
-  // Architecture related things
+template<typename base_t, typename... base_v>
+struct inherit_t<base_t, base_v...> :
+base_t::template rebind<inherit_t<base_v...>> {};  // What does ::template do here?
+
+template<typename base_t>
+struct inherit_t<base_t> : base_t {};
+////////////////
+
+/**
+ * @brief Struct holding kernel parameters will be passed in upon launch
+ * @tparam blockDim_ 1D block dimensions to launch with
+ * @tparam gridDim_ 1D grid dimensions to launch with
+ * @tparam smemBytes_ Amount of shared memory to allocate
+ */
+template<
+  unsigned int blockDim_,
+  unsigned int gridDim_,
+  unsigned int smemBytes_ = 0
+>
+struct launch_params_t {
+  enum : unsigned int {
+    blockDim = blockDim_,
+    gridDim = gridDim_,
+    smemBytes = smemBytes_
+  };
 };
 
-// Is there a better way to create arch types?
-#define NAMED_ARCH_LP_TYPE(name) \
-template<int... launch_tparams> \
-using name = struct arch_launch_params_t<launch_tparams...>;
-
-NAMED_ARCH_LP_TYPE(ampere)
-NAMED_ARCH_LP_TYPE(kepler)
-NAMED_ARCH_LP_TYPE(maxwell)
-NAMED_ARCH_LP_TYPE(pascal)
-NAMED_ARCH_LP_TYPE(volta)
-NAMED_ARCH_LP_TYPE(turing)
-
-template<typename... archLaunchParams>
-struct launch_box_t {
-  // Provide a way to get the correct params?
+// Create launch param structs for each architecture for use in the launch box struct
+// rebind type allows inherit_t to re-set the base class to each different arch_launch_param_t
+// Is there a way to make archname##_arch_t generic and only macro the type alias template?
```

Comment:
I am a bit confused about what this macro is doing. Do you wanna inherit all architectures here? Or some user selected? You could use variadic inheritance here again.

Reply:
The purpose of this macro is to create struct templates for each architecture, so their types can be set by the user (e.g. `ampere<16, 64, 2>`) when they fill in the launch params when setting the launch box. Like in the unit test I wrote.